### ISSUE TITLE
[Bug]: Failing tests have a trace with no images and with steps missing

### ISSUE THREAD

**Issue by cstlh** [microsoft/playwright#31541]
### Version

1.45.1

### Steps to reproduce

- Clone https://github.com/cstlh/playwright-trace-issue
- Run `npm -i`
- Run `npx playwright test`
- Open report
- Check test that is failing and is using baseTest.ts has a mostly empty trace and most steps are missing
  <img width="300" alt="image" src="https://github.com/microsoft/playwright/assets/4043323/8072d172-4d6f-4d20-b4e4-f663d7d0303c">

- Check that failed test that is not using baseTest.ts, but instead using @playwright/test directly has all trace information
  <img width="300" alt="image" src="https://github.com/microsoft/playwright/assets/4043323/76f5ffd7-225e-44d6-925f-393effda0d47">

- Check that passed test using baseTest.ts has full trace and steps

Additional steps:
 - Revert back to playwright 1.43.1 `npm i -D @playwright/test@1.43.1`
 - Run tests again and check that all reports have full trace

### Expected behavior

I expected that the whole trace is present in all failing tests

### Actual behavior

Traces are mostly empty in failed tests

### Additional context

I have tried the same setup in 1.43 and 1.44 and everything was working fine, this started failing for me when I tried to update the version to 1.45 and 1.45.1

### Environment

```shell
System:
    OS: macOS 14.4.1
    CPU: (8) arm64 Apple M2
    Memory: 1004.56 MB / 24.00 GB
  Binaries:
    Node: 20.11.1 - /usr/local/bin/node
    npm: 10.8.0 - /usr/local/bin/npm
    pnpm: 9.4.0 - /usr/local/bin/pnpm
  IDEs:
    VSCode: 1.90.2 - /usr/local/bin/code
  Languages:
    Bash: 3.2.57 - /bin/bash
  npmPackages:
    @playwright/test: ^1.45.1 => 1.45.1
```

**Comment by mxschmitt** [comment]
I can reproduce that failTestNoTrace.spec reproduces no trace in 1.45 but produces one in 1.44.

**Comment by rberovskis** [comment]
I was about to create a bug for the same issue myself. I am facing the same issue just without the use of base class. 

I've done some investigation and I think its even more straightforward - the issue might lie in the fact that it creates context and doesn't save it. It used to work just fine, but its no longer working. 

[I have an example repository here](https://github.com/rberovskis/pw-reporting-bug)

For me, workaround is pretty straightforward - just save Context outside of beforeAll and then close it. This works with `"@playwright/test": "^1.45",`

Basically this used to work

  ```ts
let page: Page;

  //This used to work just fine
  test.beforeAll(async ({ browser }) => {
    // Same thing - no longer works as well
    // const context = await browser.newContext();
    // page = await context.newPage();

    page = await (await browser.newContext()).newPage();
  });  

  test("get started link", async () => {
    await page.goto("https://playwright.dev/");

    // Click the get started link.
    await page.getByRole("link", { name: "Get started" }).click();

    // Expects page to have a heading with the name of Installation.
    await expect(
      page.getByRole("heading", { name: "Installationnn" })
    ).toBeVisible();
  });
```
Now I need to do this
  
```ts
  let page: Page;
  let context: Context;

  test.beforeAll(async ({ browser }) => {
    context = await browser.newContext();
    page = await context.newPage();
  });

  test("get started link", async () => {
    await page.goto("https://playwright.dev/");

    // Click the get started link.
    await page.getByRole("link", { name: "Get started" }).click();

    // Expects page to have a heading with the name of Installation.
    await expect(
      page.getByRole("heading", { name: "Installationnn" })
    ).toBeVisible();
  });
  
  
  test.afterAll("Cleanup", async () => {
    await page.close();
    await context.close();
  });
  
```

**Comment by mxschmitt** [comment]
I can verify that this bug (original reporter) is fixed in @beta and should be available in v1.45.2 soon.

**Comment by Tanmay012345** [comment]
Since this is fixed from 1.45.2 upwards, the main thing is to stay on a known-good Playwright version for trace recording and keep your own visibility across runs. One pattern that has worked well for us is:
– Let Playwright keep producing traces/screenshots as usual.
– Upload each CI run into an external reporting service (we use TestDino) so we can see, over time, which versions/configs produced incomplete traces vs full ones, instead of discovering it case by case in individual HTML reports. 

That doesn’t replace the fix here, but it makes it much easier to spot when trace data regresses again on future upgrades
